### DISCUSSION TITLE
Get typescript types info inside extension

### DISCUSSION THREAD

**Discussion by kurienzach** [microsoft/vscode-discussions#328] [Extension Development QnA] [answered]
I'm developing an extension for refactoring react code. Currently, I'm using jscodeshift to parse the code and apply modifications. However, I'm unable to find type information for expressions. Eg. Getting type information for props. What would be the best way to go about this?

I did some googling and found [this extension](https://github.com/MartinJohns/vscode-inline-types) which runs its own typescript compiler to get type info.  I think this approach is wasteful since VsCode already runs some sort of typescript type checking in the background (provided by the typescript extension I assume). Is there anyway to get the type info that is already present in Vscode?

Another hack I found is to call the `vscode.executeHoverProvider` command which provides the type info for variables. But this is very limited and does not provide info for expressions / functions.

**Comment by Drisate** [comment]
It sounds like you are looking for a way to get type information for expressions in your React code, and you are considering using the vscode.executeHoverProvider command to do so. This command can be used to get type information for variables, but it may not provide the type information you need for expressions or functions.

One option you could consider is using the TypeScript language server to get type information. The TypeScript language server provides a rich set of APIs that allow you to get type information for expressions and functions in your code. You can use the textDocument/typeDefinition request to get the definition of a symbol at a specific location in a file, which can include type information for expressions and functions.

To use the TypeScript language server, you will need to set up a connection to the server from your extension. You can use the vscode-languageclient npm package to set up the connection and communicate with the server using the Language Server Protocol.

Alternatively, you could consider using the @babel/plugin-syntax-typescript and @babel/preset-typescript packages to parse your TypeScript code and get type information. These packages can be used together with Babel to parse and transform TypeScript code, and they provide access to the type information through the AST.

**Reply by kurienzach** [reply]
Thanks @Drisate for the reply. When setting up a connecting to the typescript language server, would it create a new instance of the language server or can I connect to the instance of the server already running inside vscode?

Are there any examples of using vscode-languageclient package?

**Reply by Drisate** [reply]
When you set up a connection to the TypeScript language server from your extension, you will be connecting to the instance of the server that is already running inside Visual Studio Code. Visual Studio Code uses the TypeScript language server to provide type checking and other language features for TypeScript and JavaScript files.

```
const { LanguageClient } = require('vscode-languageclient');

// Set up the language client
const client = new LanguageClient('typescript', 'TypeScript', {
  // The command to start the server
  command: 'tsc',
  // The args to pass to the server
  args: ['--stdio']
});

// Start the client
client.start();
```

This example assumes that the TypeScript compiler (tsc) is installed on the system and available on the PATH. If you have the TypeScript compiler installed in a different location, you will need to specify the full path to the tsc executable in the command option.

Once you have set up the connection to the TypeScript language server, you can use the Language Server Protocol to send requests to the server and receive responses. For example, you can use the textDocument/typeDefinition request to get the definition of a symbol at a specific location in a file, which can include type information for expressions and functions.

**Reply by r3m0t** [reply]
@Drisate You're giving a command and arguments, that means it's starting a new process and not connecting to the existing instance.

**Reply by ArturoDent** [reply]
Sounds like a chatBot answer to me...

**Comment by r3m0t** [comment, marked as answer]
Did you take a look at the https://code.visualstudio.com/api/references/contribution-points#contributes.typescriptServerPlugins and https://github.com/RyanCavanaugh/sample-ts-plugin ? I haven't used it myself, but it looks like it lets you write code that runs in the typechecker. It looks like you need to publish the plugin part as a separate npm package.
